eth/beacon/helpers.py

```diff
@@ -360,20 +366,133 @@ def get_proposer_position(parent_block: 'BaseBeaconBlock',
     """
     Returns the proposer index in committee and the ``shard_id``.
     """
-    if len(shards_and_committees) <= 0:
-        raise ValueError("shards_and_committees should not be empty.")
-
     # `proposer_index_in_committee` th attester in `shard_and_committee`
     # is the proposer of the parent block.
-    shard_and_committee = shards_and_committees[0]
-    if len(shard_and_committee.committee) <= 0:
+    try:
+        shard_and_committee = shards_and_committees[0]
+    except IndexError:
+        raise ValueError("shards_and_committees should not be empty.")
+
+    proposer_committee_size = len(shard_and_committee.committee)
+    if proposer_committee_size <= 0:
         raise ValueError(
             "The first committee should not be empty"
         )
 
     proposer_index_in_committee = (
         parent_block.slot_number %
-        len(shard_and_committee.committee)
+        proposer_committee_size
+    )
+
+    # The index in CrystallizedState.validators
+    proposer_index = shard_and_committee.committee[proposer_index_in_committee]
+
+    return BlockCommitteesInfo(
+        proposer_index=proposer_index,
+        proposer_index_in_committee=proposer_index_in_committee,
+        proposer_shard_id=shard_and_committee.shard_id,
+        proposer_committee_size=proposer_committee_size,
+        shards_and_committees=shards_and_committees,
+    )
+
+
+#
+# Signatures and Aggregation
+#
+def create_signing_message(slot: int,
+                           parent_hashes: Iterable[Hash32],
+                           shard_id: int,
+                           shard_block_hash: Hash32,
+                           justified_slot: int) -> bytes:
+    # TODO: will be updated to hashed encoded attestation
+    return blake(
+        slot.to_bytes(8, byteorder='big') +
+        b''.join(parent_hashes) +
+        shard_id.to_bytes(2, byteorder='big') +
+        shard_block_hash +
+        justified_slot.to_bytes(8, 'big')
+    )
+
+
+def aggregate_attestation_record(last_justified_slot: int,
+                                 recent_block_hashes: Iterable[Hash32],
+                                 block: 'BaseBeaconBlock',
+                                 votes: Iterable[Tuple[int, bytes, int]],
+                                 proposer_attestation: 'AttestationRecord',
+                                 cycle_length: int) -> 'AttestationRecord':
+    """
+    Aggregate the votes.
+
+    TODO: Write tests
+    """
+    # Get signing message
+    parent_hashes = get_hashes_to_sign(
+        recent_block_hashes,
+        block,
+        cycle_length,
+    )
+    message = create_signing_message(
+        block.slot_number,
+        parent_hashes,
+        proposer_attestation.shard_id,
+        proposer_attestation.shard_block_hash,
+        last_justified_slot,
+    )
+
+    # Verify
+    # TODO: Verification is very slow, needs to compute in parallel
+    voting_sigs, committee_indices = verify_votes(message, votes)
```

Comment:
> What is going to be calling this function? This is for a block producer bundling attestations of the previous block? Note that they should be including attestations not just from the previous block but for any within the valid range (defined by parent block and CYCLE_LENGTH) that have not been included

> Note that they should be including attestations not just from the previous block but for any within the valid range (defined by parent block and CYCLE_LENGTH)

You're right. Verifying the attestations sigs *right after receiving it* makes more sense. (like an AttestationPool)
So upon receiving the Attestation:
1. Call `validate_attestation`
2. Store the attestation in the AttestationPool

When the client wants to aggregate and  the attestation pack: 
1. Check if we can reuse attestations from other block and set `oblique_parent_hashes`.
2. Aggregate the attestations that are siging for the same message.

Anything else I missed?

> We can aggregate and include attestations that are not the same as the proposer_attestation. For example, we can include votes that were for a short range fork rather than for the what the proposer submit as the head. Or some of the votes could vote for the proposer's block but for a different shard_block_hash

Right. I will rename `proposer_attestation` to something more general.

Reply:
> Right. I will rename proposer_attestation to something more general.

@djrtwo 

Second thought - I will just remove `aggregate_attestation_record` from here. It's not part of StateMachine. 😆 

But please check the correctness of the story above. Thanks!

Reply:
Looks good. The AttestationPool and who should be doing the aggregation might change depending on the network-level protocol we define for aggregation.

I would note that the AttestationPool should likely only include attestations of single signers (non-aggregated). Also that when we receive a new block (or create one) we should clean up our attestation pool, removing any of the attestations that were included in the block.

> Check if we can reuse attestations from other block
I would be clear that this is "reuse attestations from a block on a short range fork"

> Aggregate the attestations that are signing for the same message.
And can be validly included in the block (not too old, etc)

Reply:
Right, the aggregation is more of the Node or Validator's responsibility.